src/rabbit_delayed_message.erl

```diff
@@ -185,18 +187,7 @@ internal_delay_message(CurrTimer, Exchange, Delivery, Delay) ->
                        make_delay(DelayTS, Exchange, Delivery)),
     case erlang:read_timer(CurrTimer) of
         false ->
-            %% last timer expired, we set a new timer for
-            %% the next message to be delivered
-            case mnesia:dirty_first(?INDEX_TABLE_NAME) of
-                %% destructuring to prevent matching '$end_of_table'
-                #delay_key{timestamp = FirstTS} = Key
-                  when FirstTS < DelayTS ->
-                    %% there are messages that expired and need to be delivered
-                    {ok, start_timer(FirstTS - Now, Key)};
-                _ ->
-                    %% empty table or DelayTS <= FirstTS
-                    {ok, start_timer(Delay, make_key(DelayTS, Exchange))}
-            end;
+            {ok, CurrTimer};
```

Comment:
`go` is called only on application startup. But gen_server can be restarted any moment (by supervisor) without calling `go`, which is not the best approach, but it worked before this changes just because timer could be started by publishing message.
When starting a new gen_server, `timer` in state is being set to `make_ref`, which mimics expired timer. That also is not a good solution because we don't know if timer is expired already or have never been started.
Instead of adding a new state variable you can set `timer` to something meaningful for not yet started timer. Since this new version is using `timer` state field only once in `internal_delay_message` it can be checked there.